modules/optimization/src/functions/NearestReporterCoordinatesFunction.C

```diff
@@ -0,0 +1,205 @@
+    paramError("y_coord_name",
+               "Number of y coordinates (",
+               _coordy.size(),
+               ") does not match number of values (",
+               _nval,
+               ").");
+  else if (!_coordz.empty() && _coordz.size() != _nval)
+    paramError("z_coord_name",
+               "Number of z coordinates (",
+               _coordz.size(),
+               ") does not match number of values (",
+               _nval,
+               ").");
+  else if (!_coordt.empty() && _coordt.size() != _nval)
+    paramError("time_name",
+               "Number of times (",
+               _coordt.size(),
+               ") does not match number of values (",
+               _nval,
+               ").");
+
+  // Find times for each unique coordinate
+  _coord_mapping.clear();
+  for (const auto & i : make_range(_nval))
+  {
+    Point pt;
+    pt(0) = _coordx.empty() ? 0.0 : _coordx[i];
+    pt(1) = _coordy.empty() ? 0.0 : _coordy[i];
+    pt(2) = _coordz.empty() ? 0.0 : _coordz[i];
+    const Real time = _coordt.empty() ? 0.0 : _coordt[i];
+
+    std::vector<std::pair<Real, std::size_t>> * vec = nullptr;
+    for (auto & it : _coord_mapping)
+      if (pt.absolute_fuzzy_equals(it.first))
+      {
+        vec = &it.second;
+        break;
+      }
+    if (!vec)
+      vec = &_coord_mapping[pt];
+
+    vec->emplace_back(time, i);
+    std::sort(vec->begin(),
+              vec->end(),
+              [](const std::pair<Real, Real> & a, const std::pair<Real, Real> & b)
+              { return a.first < b.first; });
+  }
+}
+
+std::array<std::pair<Real, std::size_t>, 2>
+NearestReporterCoordinatesFunction::findNearestPoint(Real t, const Point & p) const
+{
+  if (_coord_mapping.empty())
+    buildCoordinateMapping();
+
+  // Make sure values is correct size
+  if (_values.size() != _nval)
+    paramError("value_name",
+               "Size of value vector (",
+               _values.size(),
+               ") does not match number of coordinates specified (",
+               _nval,
+               ").");
+
+  const auto & tval =
+      std::min_element(_coord_mapping.begin(),
+                       _coord_mapping.end(),
+                       [&p](const std::pair<Point, std::vector<std::pair<Real, std::size_t>>> & p1,
+                            const std::pair<Point, std::vector<std::pair<Real, std::size_t>>> & p2)
+                       { return (p - p1.first).norm_sq() < (p - p2.first).norm_sq(); })
+          ->second;
+
+  if (tval.size() == 1 || MooseUtils::absoluteFuzzyLessEqual(t, tval[0].first))
+    return {tval[0], tval[0]};
+  else if (MooseUtils::absoluteFuzzyGreaterEqual(t, tval.back().first))
+    return {tval.back(), tval.back()};
+  else
+    for (std::size_t ti = 1; ti < tval.size(); ++ti)
+      if (MooseUtils::absoluteFuzzyGreaterEqual(tval[ti].first, t))
+        return {tval[ti - 1], tval[ti]};
+  return std::array<std::pair<Real, std::size_t>, 2>();
```

Comment:
```suggestion
```

Reply:
This is needed to avoid compile error. Will add an error in case something went terribly wrong.

Reply:
`mooseError` is tagged `[noreturn]` so the compiler would be wrong complaining here,